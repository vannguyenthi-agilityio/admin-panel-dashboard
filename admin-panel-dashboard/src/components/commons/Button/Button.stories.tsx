import type { Meta, StoryObj } from '@storybook/react';
import Button from './';
import {
  PrevIcon,
  NextIcon,
  DeleteIcon,
  EditIcon,
  SortIcon,
  MailIcon,
  HelpIcon,
  ArrowLeftIcon,
  ArrowRightIcon,
  ArrowUpRightIcon
} from '@/components/Icons';
import { COLOR } from '@/constants';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction
const meta = {
  title: 'Components/Commons/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {},
} as Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary: Story = {
  args: {
    label: 'Cancel',
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const Secondary: Story = {
  args: {
    label: 'Create',
    buttonType: 'secondary',
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const Pagination: Story = {
  args: {
    label: '2',
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const PaginationActive: Story = {
  args: {
    label: '1',
    size: 'small',
    className: 'w-[40px] h-[40px]',
    buttonType: 'blood',
  },
};

export const PaginationPrev: Story = {
  args: {
    rounded: 'lg',
    size: 'small',
    className: 'w-[40px] h-[40px]',
    buttonType: 'disabled',
    leftIcon: <PrevIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const PaginationNext: Story = {
  args: {
    rounded: 'lg',
    size: 'small',
    className: 'w-[40px] h-[40px]',
    leftIcon: <NextIcon stroke = {COLOR.SECONDARY[200]}/>,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const ArrowRightWithIcon: Story = {
  args: {
    size: 'small',
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <ArrowRightIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const ArrowUpRightWithIcon: Story = {
  args: {
    size: 'small',
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <ArrowUpRightIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const ArrowLeftWithIcon: Story = {
  args: {
    size: 'small',
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <ArrowLeftIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const DeleteWithIcon: Story = {
  args: {
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <DeleteIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const EditWithIcon: Story = {
  args: {
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <EditIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const HelpWithIcon: Story = {
  args: {
    className: 'w-[40px] h-[40px] p-0 bg-black',
    leftIcon: <HelpIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const SortWithIcon: Story = {
  args: {
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <SortIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const MailWithIcon: Story = {
  args: {
    className: 'w-[40px] h-[40px] p-0',
    leftIcon: <MailIcon />,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};

export const Disabled: Story = {
  args: {
    label: 'Create',
    buttonType: 'disabled',
  },
};

export const SmallWithIcon: Story = {
  args: {
    size: 'small',
    label: 'New Customer',
    className: 'gap-x-2',
    buttonType: 'blood',
    leftIcon: <span>+</span>,
    onClick: () => {
      window.console.log('Button clicked!');
    },
  },
};
