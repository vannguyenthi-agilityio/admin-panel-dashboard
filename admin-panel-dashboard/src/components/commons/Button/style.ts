import { cva } from 'class-variance-authority';

/**
 * Button styles for the Button component.
 */
export const buttonStyles = cva(
  'flex flex-row gap-x-4 font-primary text-sm uppercase font-medium rounded-lg disabled:cursor-not-allowed items-center justify-center',
  {
    variants: {
      buttonType: {
        transprent: 'bg-transprent border-none hover:bg-transprent',
        primary:
          'bg-transprent text-darker border-greyish hover:bg-greyish',
        secondary:
          'bg-blood text-white border-blood hover:bg-blood',
        warning:
          'bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600',
        outline:
          'bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md',
        disabled: 'bg-disable text-disable border-disable cursor-not-allowed',
        error: 'bg-red-500 text-white border-red-500 hover:bg-red-600',
        blood: 'bg-white text-blood hover:bg-gray-100 border hover:border-blood border-blood hover:shadow-md',
      },
      size: {
        default: ['text-base'],
        small: ['text-sm'],
        large: ['text-lg'],
        xxl: ['text-2xl'],
      },
      spacing: {
        default: ['py-2', 'px-4'],
        small: ['py-1', 'px-2'],
        large: ['py-3', 'px-6'],
        xxl: ['py-4', 'px-8'],
      },
      rounded: {
        default: 'rounded-md',
        sm: 'rounded-sm',
        lg: 'rounded-lg',
        xl: 'rounded-xl',
        xxl: 'rounded-2xl',
        none: 'rounded-none',
        full: 'rounded-full',
      },
    },
    compoundVariants: [
      {
        buttonType: 'outline',
        size: 'small',
        spacing: 'default',
        rounded: 'lg',
      },
    ],
    defaultVariants: {
      buttonType: 'outline',
      size: 'small',
      rounded: 'lg',
      spacing: 'default',
    },
  }
);
