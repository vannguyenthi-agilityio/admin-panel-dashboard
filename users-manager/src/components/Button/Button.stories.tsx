import Storybook from '@storybook/react';
import { Button } from './index';

export default {
  title: 'Components/Button',
  component: Button,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    backgroundColor: { control: 'color' }
  }
} as Storybook.ComponentMeta<typeof Button>;

const Template: Storybook.ComponentStory<typeof Button> = (args) => (
  <Button label="Edit" {...args} />
);

export const Secondary = Template.bind({});
Secondary.args = {
  size: 'default',
  variant: 'colorDefault'
};

export const Transparent = Template.bind({});
Transparent.args = {
  size: 'default',
  color: 'default.red',
  bg: 'transparent',
  variant: 'border'
};

export const Cancel = Template.bind({});
Cancel.args = {
  size: 'default',
  variant: 'cancel',
  color: 'default.grey.600',
  bg: 'transparent'
};

export const Loading = Template.bind({});
Loading.args = {
  size: 'default',
  isLoading: true
};

export const Disabled = Template.bind({});
Disabled.args = {
  size: 'default',
  isDisabled: true
};
