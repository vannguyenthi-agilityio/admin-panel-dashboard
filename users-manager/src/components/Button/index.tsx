import React from 'react';
import {
  Button as ButtonChakra,
  ButtonProps as ButtonPropsChakra
} from '@chakra-ui/react';

interface ButtonProps extends ButtonPropsChakra {
  children?: React.ReactNode;
  label?: string;
  className?: string;
  isLoading?: boolean;
  isDisabled?: boolean;
  primary?: boolean;
  backgroundColor?: string;
  size?: 'default' | 'autoSize' | 'large' | 'medium' | 'small';
  variant?: 'colorDefault' | 'status' | 'transparent' | 'border' | 'muted';
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Button = ({
  size = 'default',
  children,
  label,
  backgroundColor,
  isDisabled = false,
  onClick,
  className = '',
  variant = 'colorDefault',
  ...props
}: ButtonProps) => (
  <ButtonChakra
    size={size}
    onClick={onClick}
    variant={variant}
    backgroundColor={backgroundColor}
    isDisabled={isDisabled}
    className={className}
    {...props}
  >
    {label}
    {children}
  </ButtonChakra>
);
