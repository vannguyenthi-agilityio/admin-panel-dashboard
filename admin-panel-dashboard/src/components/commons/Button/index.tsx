import { memo } from 'react';

// Helpers
import { clsxMerge } from '@/helpers';

// Types
import { IButton } from "@/types";

import { buttonStyles } from './style';

const Button = memo(({
  className,
  buttonType,
  size,
  rounded,
  label,
  rightIcon,
  spacing,
  leftIcon,
  ...props
}: IButton) => (
  <button
    className={clsxMerge(
      buttonStyles({ buttonType, size, rounded, spacing }),
      className
    )}
    type="button"
    {...props}
  >
    {leftIcon && leftIcon}
    {label && label}
    {rightIcon && rightIcon}
  </button>
));
Button.displayName = "Button";
export default Button;
