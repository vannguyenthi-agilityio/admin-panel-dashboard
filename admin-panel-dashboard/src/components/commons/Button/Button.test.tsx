import {
  render,
  screen,
  fireEvent
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from '.';

const mockOnClick = jest.fn();

describe('Button component', () => {
  it('Renders correctly', () => {
    const { container } = render(<Button>Click Me</Button>);

    expect(container.firstChild).toHaveClass('bg-white');
    expect(container.firstChild).toHaveClass('text-darker');
    expect(container.firstChild).toHaveClass('rounded-lg');
    expect(container.firstChild).toHaveClass('px-4');
  });

  it('Renders correctly with left icon', () => {
    const { container } = render(
      <Button leftIcon={<span>ğŸ‘ˆ</span>}>Click Me</Button>
    );

    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('ğŸ‘ˆ');
  });

  it('Renders correctly with right icon', () => {
    const { container } = render(
      <Button rightIcon={<span>ğŸ‘‰</span>}>Click Me</Button>
    );

    expect(container).toMatchSnapshot();
    expect(container).toHaveTextContent('ğŸ‘‰');
  });

  it('Renders correctly with both icons', () => {
    const { container } = render(
      <Button leftIcon={<span>ğŸ‘ˆ</span>} rightIcon={<span>ğŸ‘‰</span>} />
    );

    expect(container).toHaveTextContent('ğŸ‘ˆ');
    expect(container).toHaveTextContent('ğŸ‘‰');
  });

  it('Renders correctly with label', () => {
    const { container } = render(<Button label="Click Me" />);

    expect(container).toHaveTextContent('Click Me');
  });

  it('Renders correctly with label and left icon', () => {
    const { container } = render(
      <Button
        label="Click Me"
        leftIcon={<span>ğŸ‘ˆ</span>}
        rightIcon={<span>ğŸ‘‰</span>}
      />
    );

    expect(container).toHaveTextContent('Click Me');
    expect(container).toHaveTextContent('ğŸ‘ˆ');
    expect(container).toHaveTextContent('ğŸ‘‰');
  });

  it('Render correctly with size prop', () => {
    const { container } = render(<Button size="default" />);

    const { container: containerSmall } = render(<Button size="small" />);

    const { container: containerLarge } = render(<Button size="large" />);

    const { container: containerXXL } = render(<Button size="xxl" />);

    expect(container.firstChild).toHaveClass('text-base');
    expect(containerSmall.firstChild).toHaveClass('text-sm');
    expect(containerLarge.firstChild).toHaveClass('text-lg');
    expect(containerXXL.firstChild).toHaveClass('text-2xl');
  });

  it('Render correctly with rounded prop', () => {
    const { container } = render(<Button rounded="full" />);

    const { container: containerLg } = render(<Button rounded="lg" />);

    const { container: containerNone } = render(<Button rounded="none" />);

    const { container: containerXL } = render(<Button rounded="xl" />);

    const { container: containerXXL } = render(<Button rounded="xxl" />);

    const { container: containerDefault } = render(
      <Button rounded="default" />
    );

    const { container: containerSM } = render(<Button rounded="sm" />);

    expect(container).toMatchSnapshot();
    expect(containerLg).toMatchSnapshot();
    expect(containerNone).toMatchSnapshot();
    expect(containerXL).toMatchSnapshot();
    expect(containerXXL).toMatchSnapshot();
    expect(containerDefault).toMatchSnapshot();
    expect(containerSM).toMatchSnapshot();

    expect(container.firstChild).toHaveClass('rounded-full');
    expect(containerLg.firstChild).toHaveClass('rounded-lg');
    expect(containerNone.firstChild).toHaveClass('rounded-none');
    expect(containerXL.firstChild).toHaveClass('rounded-xl');
    expect(containerXXL.firstChild).toHaveClass('rounded-2xl');
    expect(containerDefault.firstChild).toHaveClass('rounded-md');
    expect(containerSM.firstChild).toHaveClass('rounded-sm');
  });

  it('Render correctly with buttonType prop', () => {

    const { container } = render(<Button buttonType="secondary" />);

    const { container: containerWarning } = render(
      <Button buttonType="warning" />
    );

    const { container: containerError } = render(<Button buttonType="error" />);

    expect(container).toMatchSnapshot();
    expect(containerWarning).toMatchSnapshot();
    expect(containerError).toMatchSnapshot();

    expect(containerWarning.firstChild).toHaveClass(
      'bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600'
    );

    expect(containerError.firstChild).toHaveClass(
      'bg-red-500 text-white border-red-500 hover:bg-red-600'
    );
  });

  it('Calls onClick handler when clicked', () => {
    render(<Button onClick={mockOnClick} label='Clickable Button'/>);
    const buttonElement = screen.getByRole('button', { name: /Clickable Button/i });
    fireEvent.click(buttonElement);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });
});
