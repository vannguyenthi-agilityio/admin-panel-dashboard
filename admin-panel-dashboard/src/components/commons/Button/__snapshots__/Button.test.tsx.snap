// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Render correctly with buttonType prop 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-blood text-white border-blood hover:bg-blood text-sm py-2 px-4 rounded-lg"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with buttonType prop 2`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-yellow-500 text-white border-yellow-500 hover:bg-yellow-600 text-sm py-2 px-4 rounded-lg"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with buttonType prop 3`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-red-500 text-white border-red-500 hover:bg-red-600 text-sm py-2 px-4 rounded-lg"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-full"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 2`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-lg"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 3`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-none"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 4`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-xl"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 5`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-2xl"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 6`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-md"
    type="button"
  />
</div>
`;

exports[`Button component Render correctly with rounded prop 7`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-sm"
    type="button"
  />
</div>
`;

exports[`Button component Renders correctly with left icon 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-lg"
    type="button"
  >
    <span>
      ðŸ‘ˆ
    </span>
  </button>
</div>
`;

exports[`Button component Renders correctly with right icon 1`] = `
<div>
  <button
    class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-lg"
    type="button"
  >
    <span>
      ðŸ‘‰
    </span>
  </button>
</div>
`;
